### PR TITLE
BUG: Fix combine_first converts other columns type into floats unexpectedly

### LINKED ISSUE #20699: Combine_first method converts booleans into floats
#### Works as expected:

```python
>>> df1 = pd.DataFrame()
>>> df2 = pd.DataFrame({'isBool': [True]})
>>> df1.combine_first(df2)
   isBool
0    True
```
#### Does not work as expected:

```python
>>> df1 = pd.DataFrame({'isInt': [1]})
>>> df2 = pd.DataFrame({'isBool': [True]})
>>> df1.combine_first(df2)
   isBool  isInt
0     1.0      1
```
#### Problem description
Would expect Pandas to preserve the bool dtype (or at least the behavior should differ for the case where df1 is empty or not).

See also: https://stackoverflow.com/questions/39103144/pandas-dataframe-combine-first-method-converts-boolean-in-floats
And related: https://stackoverflow.com/questions/15349795/pandas-dataframe-combine-first-and-update-methods-have-strange-behavior
And: https://github.com/pandas-dev/pandas/issues/3016

Seems to me some of the issues where solved but this case may have escaped(?)

#### Output of ``pd.show_versions()``:
<details>
commit: None
python: 3.6.3.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 158 Stepping 9, GenuineIntel
byteorder: little
LC_ALL: None
LANG: en
LOCALE: None.None

pandas: 0.21.1
pytest: 3.5.0
pip: 9.0.3
setuptools: 38.5.2
Cython: None
numpy: 1.13.3
scipy: 1.0.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.7.1
patsy: None
dateutil: 2.7.0
pytz: 2018.3
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.1.1
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 1.0.1
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by Licht-T**
- [x] closes #20699
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -4870,14 +4870,24 @@ def combine(self, other, func, fill_value=None, overwrite=True):
                 series[this_mask] = fill_value
                 otherSeries[other_mask] = fill_value
 
-            # if we have different dtypes, possibly promote
-            new_dtype = this_dtype
-            if not is_dtype_equal(this_dtype, other_dtype):
```
Comment (jreback) [inline pandas/core/frame.py]: any way to simplify this, adding this condition is a bit odd; if these are not the same dtype then we will astype anyhow, why does adding this extra check help here?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 1 above)
Comment (Licht-T) [reply pandas/core/frame.py]: Ah! This line is none-sense because `find_common_type` exists in the next line!

**Comment by jreback** [comment]
can you add a whatsnew note. also rebase on master; I cleaned some older code out of here (shouldn't impact but just to be sure)

**Comment by jorisvandenbossche** [comment]
@Licht-T Do you have time to update this?

**Review by jreback — CHANGES_REQUESTED** [review]
can you update and add a whatnew (0.23.2) is ok

**Comment by jreback** [comment]
can you rebase / update?

@gfyoung if not, can you fix the note

**Comment by gfyoung** [comment]
@jreback : Comments addressed, and all is rebased and green.  PTAL.

**Comment by gfyoung** [comment]
@jreback : Comments addressed, and all is rebased and green. PTAL.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py [hunk 2]
```diff
@@ -5072,9 +5072,17 @@ def combine(self, other, func, fill_value=None, overwrite=True):
...
+            if col not in self.columns:
+                # If self DataFrame does not have col in other DataFrame,
+                # try to promote series, which is all NaN, as other_dtype.
+                new_dtype = other_dtype
+                try:
+                    series = series.astype(new_dtype)
```
[the remark this answers is not in the document]
Comment (gfyoung) [reply pandas/core/frame.py]: Done.

File: pandas/tests/frame/test_combine_concat.py [hunk 3]
```diff
@@ -750,6 +750,16 @@ def test_combine_first_int(self):
         tm.assert_frame_equal(res, df1)
         assert res['a'].dtype == 'int64'
 
+    def test_combine_first_with_asymmetric_other(self):
```
[the remark this answers is not in the document]
Comment (gfyoung) [reply pandas/tests/frame/test_combine_concat.py]: Done.

**Comment by gfyoung** [comment]
@jreback : Comments addressed, and all is rebased and green. PTAL.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @Licht-T (and @gfyoung )!

note that this may close other combine / combine_first issues. maybe have a look if you have a chance.

**Pull request merged**